presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaFilterManager.java

```diff
@@ -0,0 +1,319 @@
+                    .filter(partitionInfo -> finalPartitionIdsFiltered.contains((long) partitionInfo.partition()))
+                    .collect(toImmutableList());
+            return new KafkaFilteringResult(partitionFilteredInfos, partitionBeginOffsets, partitionEndOffsets);
+        }
+        return new KafkaFilteringResult(partitionInfos, partitionBeginOffsets, partitionEndOffsets);
+    }
+
+    private boolean isTimestampUpperBoundPushdownEnabled(ConnectorSession session, String topic)
+    {
+        try (AdminClient adminClient = adminFactory.create()) {
+            ConfigResource topicResource = new ConfigResource(ConfigResource.Type.TOPIC, topic);
+
+            DescribeConfigsResult describeResult = adminClient.describeConfigs(Collections.singleton(topicResource));
+            Map<ConfigResource, Config> config = describeResult.all().get();
+
+            if (config != null) {
+                Config c = config.get(topicResource);
+                String timestampType = c.get(TOPIC_CONFIG_TIMESTAMP_KEY).value();
+                if (TOPIC_CONFIG_TIMESTAMP_VALUE_LOG_APPEND_TIME.equals(timestampType)) {
+                    return true;
+                }
+            }
+        }
+        catch (Exception e) {
+            throw new PrestoException(KAFKA_SPLIT_ERROR, format("Failed to get topic(%s) config", topic), e);
+        }
+        return KafkaSessionProperties.isTimestampUpperBoundPushdownEnabled(session);
+    }
+
+    private static Optional<Long> findOffsetsByTimestamp(KafkaConsumer<byte[], byte[]> kafkaConsumer, TopicPartition topicPartition, long timestamp)
+    {
+        try {
+            if ((timestamp + "").length() == 16) {
+                timestamp /= 1000;
+            }
+            Map<TopicPartition, OffsetAndTimestamp> topicPartitionOffsets = kafkaConsumer.offsetsForTimes(ImmutableMap.of(topicPartition, timestamp));
+            return Optional.ofNullable(getOnlyElement(topicPartitionOffsets.values(), null)).map(OffsetAndTimestamp::offset);
+        }
+        catch (Exception e) {
+            return Optional.empty();
+        }
+    }
+
+    private static Map<TopicPartition, Long> calcTopicPartitionBeginOffsetMap(Map<TopicPartition, Long> partitionBeginOffsets,
+                                                                             Function<TopicPartition, Optional<Long>> computeOffset)
+    {
+        final Map<TopicPartition, Long> partitionFilteredBeginOffsets = new HashMap<>();
+        partitionBeginOffsets.forEach((partition, partitionIndex) -> {
+            Optional<Long> filterIndex = computeOffset.apply(partition);
+            partitionFilteredBeginOffsets.put(partition, filterIndex.map(index -> Long.max(partitionIndex, index)).orElse(partitionIndex));
+        });
+        return partitionFilteredBeginOffsets;
+    }
+
+    private static Map<TopicPartition, Long> calcTopicPartitionEndOffsetMap(Map<TopicPartition, Long> partitionEndOffsets,
```

Comment:
Are EndOffsets inclusive or exclusive?
You are calling `calcTopicPartitionEndOffsetMap` as if they were exclusive. Is that the case?

Reply:
Yes，endOffsets is exclusive. Actually it is hard judging timestamp range as judging partition_offset